File: selectionSort/selectionSort.c
/*
  選択ソート
*/

#include <stdio.h>

void trace(int array[],int length) {
  for(int i=0; i < length; i++) {
    printf("%d ", array[i]);
  }
  printf("\n");
}

void swap(int *a, int *b) {
  int temp = *a;
  *a = *b;
  *b = temp;
}

void selectionSort(int array[],int length) {
  for(int i = 0; i < length; i++) {
    int minj = i;
    for(int j = i; j < length; j++) {
      if(array[j] < array[minj]) minj = j;
    }
    swap(&array[i],&array[minj]);
    trace(array, length);
  }
}

int main(void) {
  int N, i, j;
  int A[100];

  printf("length : "); scanf("%d", &N);
  for(i=0; i<N; i++) scanf("%d", &A[i]);

  trace(A,N);
  selectionSort(A,N);

  return 0;
}

File: selectionSort/selectionSort.lua

function trace(array, length)
   for idx = 1, length, 1 do
      io.write(array[idx]);
      io.write(" ")
   end
   print("")
end

function selectionSort(array, length)
   for idx1 = 1, length-1, 1 do
      mindex = idx1
      for idx2 = idx1+1, length, 1 do
         if array[idx2] < array[mindex] then
            mindex = idx2
         end
      end
      array[idx1], array[mindex] = array[mindex], array[idx1]
      trace(array, length)
   end
end



--val1, val2 = swap(val1, val2)
array = {9,3,6,8,5,4,7,2,1}

selectionSort(array, 9)

